src/runtime_src/core/common/api/xrt_xclbin.cpp

```diff
@@ -38,126 +41,919 @@
+      m_axlf_sections.emplace(kind, std::make_pair(section_data, section_header.m_sectionSize));
+    }
+
+    /* If group topology or group connectivity doesn't exists then use the
+     * mem topology or connectivity respectively section for the same.
+     */
+    if (m_axlf_sections.find(ASK_GROUP_TOPOLOGY) == m_axlf_sections.end() &&
+        m_axlf_sections.find(MEM_TOPOLOGY) != m_axlf_sections.end()) {
+      m_axlf_sections[ASK_GROUP_TOPOLOGY] = m_axlf_sections[MEM_TOPOLOGY];
+    }
+    if (m_axlf_sections.find(ASK_GROUP_CONNECTIVITY) == m_axlf_sections.end() &&
+        m_axlf_sections.find(CONNECTIVITY) != m_axlf_sections.end()) {
+      m_axlf_sections[ASK_GROUP_CONNECTIVITY] = m_axlf_sections[CONNECTIVITY];
+    }
   }
 
-  const std::vector<char>&
-  get_data() const
+  explicit xclbin_full(const axlf* top)
+  {
+    if (strncmp(get_top()->m_magic, "xclbin2", 7))
+      throw std::runtime_error("Invalid xclbin");
+
+    auto length = top->m_header.m_length;
+    m_axlf.resize(length);
+    auto top_c = reinterpret_cast<const char*>(top);
+    std::copy(top_c,top_c+length,m_axlf.begin());
+
+    m_uuid = uuid(get_top()->m_header.uuid);
+
+    auto vbnv = reinterpret_cast<const char*>(get_top()->m_header.m_platformVBNV);
+    m_vbnv = std::string(vbnv, strnlen(vbnv, 64));
+
+    // Build Map
+    for (const auto& section_header : xrt_core::xclbin_int::axlf_header_container(get_top())) {
+      auto kind = static_cast<axlf_section_kind>(section_header.m_sectionKind);
+      auto section_data = reinterpret_cast<const char*>(get_top()) + section_header.m_sectionOffset;
+      m_axlf_sections.emplace(kind, std::make_pair(section_data, section_header.m_sectionSize));
+    }
+
+    /* If group topology or group connectivity doesn't exists then use the
+     * mem topology or connectivity respectively section for the same.
+     */
+    if (m_axlf_sections.find(ASK_GROUP_TOPOLOGY) == m_axlf_sections.end() &&
+        m_axlf_sections.find(MEM_TOPOLOGY) != m_axlf_sections.end()) {
+      m_axlf_sections[ASK_GROUP_TOPOLOGY] = m_axlf_sections[MEM_TOPOLOGY];
+    }
+    if (m_axlf_sections.find(ASK_GROUP_CONNECTIVITY) == m_axlf_sections.end() &&
+        m_axlf_sections.find(CONNECTIVITY) != m_axlf_sections.end()) {
+      m_axlf_sections[ASK_GROUP_CONNECTIVITY] = m_axlf_sections[CONNECTIVITY];
+    }
+  }
+
+  const std::vector<char>& get_data() const override
   {
     return m_axlf;
   }
+
+  std::pair<const char*, size_t>
+  get_axlf_section(axlf_section_kind section, const uuid& xclbin_id) const override
+  {
+    if (xclbin_id && xclbin_id != m_uuid)
+      throw std::runtime_error("xclbin id mismatch");
+    auto itr = m_axlf_sections.find(section);
+    return itr != m_axlf_sections.end()
+           ? (*itr).second
+           : std::make_pair(nullptr, size_t(0));
+  }
+
+protected:
+  std::vector<char> m_axlf;
+  std::map<axlf_section_kind, std::pair<const char*, size_t>> m_axlf_sections;
+
+  axlf*
```

Comment:
Return type should be `const axlf*`  Caller should not be allowed to modify the axlf returned.